cpp/src/arrow/array/util.cc

```diff
@@ -74,6 +74,177 @@ class ArrayDataWrapper {
+    for (int64_t i = 0; i < length; i++) {
+#if ARROW_LITTLE_ENDIAN
+      buffer[i] = BitUtil::FromBigEndian(buffer[i]);
+#else
+      buffer[i] = BitUtil::FromLittleEndian(buffer[i]);
+#endif
+    }
+    return Status::OK();
+  }
+
+  Status Visit(const Decimal128Type& type) {
+    auto buffer = const_cast<uint64_t*>(
+        reinterpret_cast<const uint64_t*>(data_->buffers[1]->data()));
+    int64_t length = length_;
+    for (int64_t i = 0; i < length; i++) {
+      uint64_t tmp;
+      auto idx = i * 2;
+#if ARROW_LITTLE_ENDIAN
+      tmp = BitUtil::FromBigEndian(buffer[idx]);
+      buffer[idx] = BitUtil::FromBigEndian(buffer[idx + 1]);
+      buffer[idx + 1] = tmp;
+#else
+      tmp = BitUtil::FromLittleEndian(buffer[idx]);
+      buffer[idx] = BitUtil::FromLittleEndian(buffer[idx + 1]);
+      buffer[idx + 1] = tmp;
+#endif
+    }
+    return Status::OK();
+  }
+
+  Status Visit(const DayTimeIntervalType& type) {
+    auto buffer = const_cast<uint32_t*>(
+        reinterpret_cast<const uint32_t*>(data_->buffers[1]->data()));
+    int64_t length = length_;
+    for (int64_t i = 0; i < length; i++) {
+      auto idx = i * 2;
+#if ARROW_LITTLE_ENDIAN
+      buffer[idx] = BitUtil::FromBigEndian(buffer[idx]);
+      buffer[idx + 1] = BitUtil::FromBigEndian(buffer[idx + 1]);
+#else
+      buffer[idx] = BitUtil::FromLittleEndian(buffer[idx]);
+      buffer[idx + 1] = BitUtil::FromLittleEndian(buffer[idx + 1]);
+#endif
+    }
+    return Status::OK();
+  }
+
+  Status Visit(const NullType& type) { return Status::OK(); }
+  Status Visit(const BooleanType& type) { return Status::OK(); }
+  Status Visit(const Int8Type& type) { return Status::OK(); }
+  Status Visit(const UInt8Type& type) { return Status::OK(); }
+  Status Visit(const FixedSizeBinaryType& type) { return Status::OK(); }
+
+  Status Visit(const StringType& type) {
+    RETURN_NOT_OK(SwapSmallOffset());
+    return Status::OK();
+  }
+  Status Visit(const LargeStringType& type) {
+    RETURN_NOT_OK(SwapLargeOffset());
+    return Status::OK();
+  }
+  Status Visit(const BinaryType& type) {
+    RETURN_NOT_OK(SwapSmallOffset());
+    return Status::OK();
+  }
+  Status Visit(const LargeBinaryType& type) {
+    RETURN_NOT_OK(SwapLargeOffset());
+    return Status::OK();
+  }
+
+  Status Visit(const ListType& type) {
+    RETURN_NOT_OK(SwapSmallOffset());
+    RETURN_NOT_OK(SwapChildren(type.fields()));
+    return Status::OK();
+  }
+  Status Visit(const LargeListType& type) {
+    RETURN_NOT_OK(SwapLargeOffset());
+    RETURN_NOT_OK(SwapChildren(type.fields()));
+    return Status::OK();
+  }
+  Status Visit(const FixedSizeListType& type) {
+    RETURN_NOT_OK(SwapChildren(type.fields()));
+    return Status::OK();
+  }
+
+  Status Visit(const MapType& type) {
+    RETURN_NOT_OK(SwapSmallOffset());
+    RETURN_NOT_OK(SwapChildren(type.fields()));
+    return Status::OK();
+  }
+
+  Status Visit(const StructType& type) {
+    RETURN_NOT_OK(SwapChildren(type.fields()));
```

Comment:
`SwapChildren` can be called in all cases instead of calling it specifically for some types.